reentrancy-counter/src/agent.spec.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  Finding,
+  HandleTransaction,
+  createTransactionEvent,
+  Trace,
+  Transaction,
+  Receipt,
+  Block,
+  TransactionEvent,
+  TraceAction,
+} from "forta-agent";
+import agent from "./agent";
+import { createFinding, reentracyLevel } from "./agent.utils";
+
+const createTrace = (to: string, stack: number[]): Trace => {
+  return {
+    traceAddress: stack,
+    action: {
+      to: to,
+    } as TraceAction,
+  } as Trace;
+};
+
+const createTxEvent = (traces: Trace[]) =>
+  createTransactionEvent({
+    transaction: {} as Transaction,
+    receipt: {} as Receipt,
+    block: {} as Block,
+    traces: traces,
+  });
+
+
+describe("Reentrancy counter agent tests suit", () => {
+  const handleTransaction: HandleTransaction = agent.handleTransaction;
+
+  describe("handleTransaction", () => {
+    it("Should return empty findinds if no traces provided", async () => {
+      const tx: TransactionEvent = createTxEvent([]);
+      const findings: Finding[] = await handleTransaction(tx);
+      expect(findings).toStrictEqual([]);
+    });
+
+    it("Should return empty findinds if no repetition detected", async () => {
+      const tx: TransactionEvent = createTxEvent([
+        createTrace("0x0", []),
+        createTrace("0x1", [0]),
+        createTrace("0x2", [0, 0]),
+        createTrace("0x3", [0, 0, 0]),
+      ]);
+      const findings: Finding[] = await handleTransaction(tx);
+      expect(findings).toStrictEqual([]);
+    });
+
+    it("Should ignore non reentrant calls", async () => {
+      const tx: TransactionEvent = createTxEvent([
+        createTrace("0x0", []),
+        createTrace("0x1", [0]),
+        createTrace("0x1", [1]),
+        createTrace("0x2", [1, 0]),
+        createTrace("0x2", [1, 1]),
+        createTrace("0x2", [1, 2]),
+        createTrace("0x2", [1, 3]),
+        createTrace("0x1", [2]),
+        createTrace("0x1", [3]),
+        createTrace("0x1", [4]),
+        createTrace("0x1", [5]),
+        createTrace("0x1", [6]),
+        createTrace("0x1", [7]),
+      ]);
+      const findings: Finding[] = await handleTransaction(tx);
+      expect(findings).toStrictEqual([]);
+    });
+
+    it("Should detect different levels of reentrancy", async () => {
+      // 0x0, 0x1, 0x3, 0x5, 0x6 called less than 3 times
+      // 0x2 called 3 times
+      // 0x4 called 5 times
+      const tx: TransactionEvent = createTxEvent([
+        createTrace("0x0", []),
+        createTrace("0x1", [0]),
+        createTrace("0x2", [0, 0]),
+        createTrace("0x3", [0, 0, 0]),
+        createTrace("0x2", [0, 0, 0, 0]),
+        createTrace("0x3", [0, 0, 0, 0, 0]),
+        createTrace("0x2", [0, 0, 0, 0, 0, 0]),
+        createTrace("0x4", [0, 1]),
+        createTrace("0x5", [0, 1, 0]),
+        createTrace("0x6", [0, 1, 0, 0]),
+        createTrace("0x4", [0, 1, 0, 0, 0]),
```

Comment:
can you add a small diag showing the loops/flow, a link to some diag would be great.

Reply:
Check the link above an tell me if the diag is still needed.

Reply:
Yes. I read the link. Would be better if you can provide a trace similar to in the example.

Reply:
A
  CALLs B
    CALLs G
  CALLs C
    CALLs G

Reply:
Well, I think that the calls all self explanatory (the stack is been showed), but I am going to add some indentation in the calls in order to simulate the approach.

Reply:
The indentation looks very bad, then some comments added int the test (note that calls have the same shape of the stacks). I hope it looks clearer for you now.